python/pyarrow/table.pxi

```diff
@@ -1787,6 +1941,35 @@ cdef class _Tabular(_PandasConvertible):
         """
         return _pc().take(self, indices)
 
+    def to_pydict(self):
+        raise NotImplementedError
```

Comment:
FWIW, looking at the current `to_pydict` implementations, while they are not exactly the same right now, I think it should be quite easy to rewrite it slightly so that it works for both table and record batch (in the end you iterate over the number of columns, and get each field/column and to_pylist on that)

Reply:
I'll go ahead and include this change! I'll check some basic timing just to see if anything major happens with performance.

Reply:
I ran the following code in IPython and received the results below. No performance difference found in my basic testing.

```
import random
import string
import pyarrow as pa

arr0 = [random.randint(0, 2_000_000) for _ in range(1_000_000)]
arr1 = [''.join(random.choices(string.ascii_uppercase + string.digits, k=10)) for _ in range(1_000_000)]
table = pa.table([arr0, arr1], ["c0", "c1"])
```
```%timeit table.to_pydict()```

Before
`2.03 s ± 7.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`
After
`1.99 s ± 7.99 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)`